asserts/model.go

```diff
@@ -74,55 +368,79 @@ func (mod *Model) Architecture() string {
 	return mod.HeaderString("architecture")
 }
 
-// snapWithTrack represents a snap that includes optional track
-// information like `snapName=trackName`
-type snapWithTrack string
-
-func (s snapWithTrack) Snap() string {
-	return strings.SplitN(string(s), "=", 2)[0]
-}
-
-func (s snapWithTrack) Track() string {
-	l := strings.SplitN(string(s), "=", 2)
-	if len(l) > 1 {
-		return l[1]
-	}
-	return ""
+// GadgetSnap returns the details of the gadget snap the model uses.
+func (mod *Model) GadgetSnap() *ModelSnap {
+	return mod.gadgetSnap
 }
 
 // Gadget returns the gadget snap the model uses.
 func (mod *Model) Gadget() string {
-	return snapWithTrack(mod.HeaderString("gadget")).Snap()
+	if mod.gadgetSnap == nil {
+		return ""
+	}
+	return mod.gadgetSnap.Name
 }
 
 // GadgetTrack returns the gadget track the model uses.
+// XXX this should go away
 func (mod *Model) GadgetTrack() string {
-	return snapWithTrack(mod.HeaderString("gadget")).Track()
+	if mod.gadgetSnap == nil {
+		return ""
+	}
+	return mod.gadgetSnap.Track
+}
+
+// KernelSnap returns the details of the kernel snap the model uses.
+func (mod *Model) KernelSnap() *ModelSnap {
+	return mod.kernelSnap
 }
 
 // Kernel returns the kernel snap the model uses.
+// XXX this should go away
 func (mod *Model) Kernel() string {
-	return snapWithTrack(mod.HeaderString("kernel")).Snap()
+	if mod.kernelSnap == nil {
+		return ""
+	}
+	return mod.kernelSnap.Name
 }
 
 // KernelTrack returns the kernel track the model uses.
+// XXX this should go away
 func (mod *Model) KernelTrack() string {
-	return snapWithTrack(mod.HeaderString("kernel")).Track()
+	if mod.kernelSnap == nil {
+		return ""
+	}
+	return mod.kernelSnap.Track
 }
 
 // Base returns the base snap the model uses.
 func (mod *Model) Base() string {
 	return mod.HeaderString("base")
 }
 
+// BaseSnap returns the details of the base snap the model uses.
+func (mod *Model) BaseSnap() *ModelSnap {
+	return mod.baseSnap
+}
+
 // Store returns the snap store the model uses.
 func (mod *Model) Store() string {
 	return mod.HeaderString("store")
 }
 
-// RequiredSnaps returns the snaps that must be installed at all times and cannot be removed for this model.
-func (mod *Model) RequiredSnaps() []string {
-	return mod.requiredSnaps
+// RequiredSnaps returns the snaps that must be installed at all times and cannot be removed for this model, excluding the boot snaps (gadget, kernel, boot base).
+func (mod *Model) RequiredSnaps() []naming.SnapRef {
```

Comment:
Given there are two similarly named functions (`RequiredSnaps` and `AllRequiredSnaps`) perhaps we could find a name that makes it more obvious as to what the difference is. Could this one be rightfully called `RequiredAppSnaps`?

Reply:
open to suggestions, app is not a good discriminator though because there can be bases there

Reply:
my current idea, is to switch the terminology in the doc strings to the one in firstboot and call gadget/kernel/base essential snaps,  so these would be:

RequiredWithEssentialSnaps
RequiredNoEssentialSnaps